File: app/src/main/java/com/foxtrader/app/domain/usecase/ai/agents/LitAgent.kt
package com.foxtrader.app.domain.usecase.ai.agents

import com.foxtrader.app.domain.model.AgentContext
import com.foxtrader.app.domain.model.AgentInsight
import com.foxtrader.app.domain.model.AgentName
import com.foxtrader.app.domain.model.AgentOutput
import com.foxtrader.app.domain.model.AgentStatus
import com.foxtrader.app.domain.model.Bias
import com.foxtrader.app.domain.model.Direction
import com.foxtrader.app.domain.usecase.ai.TradingAgent
import com.foxtrader.app.domain.usecase.signalintel.ConfirmedBarPolicy
import com.foxtrader.app.domain.usecase.signalintel.LitEngine
import javax.inject.Inject

/**
 * Canonical LiT (Liquidity Inducement Theory) agent.
 *
 * This agent is deliberately fail-closed. It may emit execution-grade LiT
 * evidence only when the repository's canonical [LitEngine] validates the full
 * confirmed-bar sequence. Generic upstream SMC/ICT sweeps, BOS/CHOCH events,
 * direct equal-high/equal-low heuristics, or SMT divergence are not allowed to
 * resurrect a setup rejected by [LitEngine]. Those remain independent evidence
 * in their own engines/agents and must not be double-counted as LiT.
 */
class LitAgent @Inject constructor(
    private val litEngine: LitEngine = LitEngine(
        smcDetector = com.foxtrader.app.domain.usecase.smc.SmcDetector(),
        analyzeStructure = com.foxtrader.app.domain.usecase.AnalyzeMarketStructureUseCase(),
        displacementDetector = com.foxtrader.app.domain.usecase.litx.DisplacementDetector(),
        premiumDiscount = com.foxtrader.app.domain.usecase.litx.PremiumDiscountCalculator(),
    ),
) : TradingAgent {

    override val name = AgentName.LIT
    override val description = "Canonical Liquidity Inducement Theory — validated confirmed-bar LiT sequence only."
    override val version = "2.0.0"

    override fun analyze(context: AgentContext): AgentOutput {
        val start = System.nanoTime()
        val confirmedIndex = ConfirmedBarPolicy.latestConfirmedIndex(
            context.candles,
            context.timeframe,
            System.currentTimeMillis(),
        )
        val candles = if (confirmedIndex >= 0) {
            context.candles.subList(0, confirmedIndex + 1)
        } else {
            emptyList()
        }

        if (candles.size < MIN_CANDLES) {
            return neutralOutput(name, "Insufficient confirmed data for canonical LiT analysis.", start)
        }

        val analysis = litEngine.analyze(
            symbol = context.symbol,
            timeframe = context.timeframe,
            candles = candles,
        )
        val signal = analysis.signal
            ?: return neutralOutput(
                name,
                "Canonical LiT rejected the current setup at ${analysis.stage}; no fallback evidence may override it.",
                start,
            )

        // Defense in depth: a canonical signal must be stamped on an actually
        // confirmed bar inside the exact prefix supplied to the engine.
        if (
            signal.confirmationIndex !in candles.indices ||
            signal.confirmationIndex != candles.lastIndex ||
            signal.timestamp != candles[signal.confirmationIndex].timestamp
        ) {
            return neutralOutput(
                name,
                "Canonical LiT signal failed confirmation-boundary validation.",
                start,
            )
        }

        val insight = AgentInsight(
            id = "${name}-CANONICAL-${signal.confirmationIndex}-${signal.direction}",
            agentName = name,
            type = "LIT_CANONICAL_VALIDATED",
            direction = signal.direction,
            confidence = signal.confidence.toDouble(),
            price = signal.entry,
            timestamp = signal.timestamp,
            barIndex = signal.confirmationIndex,
            detail = signal.rationale,
            weight = CANONICAL_WEIGHT,
            tags = signal.confirmations + listOf(
                "LIT",
                "CANONICAL",
                "SEQUENCE_VALIDATED",
                "NON_REPAINT",
                "NO_FALLBACK",
            ),
        )

        val bias = when (signal.direction) {
            Direction.BULLISH -> Bias.BULLISH
            Direction.BEARISH -> Bias.BEARISH
        }

        return AgentOutput(
            agentName = name,
            status = AgentStatus.COMPLETE,
            bias = bias,
            confidence = signal.confidence.toDouble(),
            insights = listOf(insight),
            narrative = "LiT: canonical validated sequence at confirmed bar ${signal.confirmationIndex}. Bias $bias.",
            processingTimeMs = elapsedMs(start),
            // Output transport time may be wall clock; the executable insight
            // itself is replay-stamped to the confirmed market candle above.
            timestamp = System.currentTimeMillis(),
        )
    }

    private companion object {
        const val MIN_CANDLES = 60
        const val CANONICAL_WEIGHT = 2.8
    }
}

File: src/agents/impl/lit-agent.ts
// ============================================================================
// AGENT 4: LIT AGENT
// Liquidity Trap, Inducement, Institutional Entry, Confirmation, Target
// ============================================================================

import { TradingAgent, AgentContext, AgentOutput, AgentInsight, AgentName } from '../types';
import { Bias, Direction } from '../../core/types';
import { LITTradingAnalyzer } from '../../modules/lit-trading';

let seq = 0;

export class LITAgent implements TradingAgent {
  readonly name: AgentName = 'LIT';
  readonly description = 'Liquidity Trap, Inducement, Institutional Entry, Confirmation Logic, Target Logic';
  readonly version = '1.0.0';
  private lit = new LITTradingAnalyzer();

  analyze(context: AgentContext): AgentOutput {
    const start = performance.now();
    const { candles } = context;
    const insights: AgentInsight[] = [];

    if (candles.length < 30) return this.empty(start);

    // Get structure/liquidity info from previous agents
    const prevStructure = context.previousOutputs?.get('MARKET_STRUCTURE');
    const prevSM = context.previousOutputs?.get('SMART_MONEY');

    // Extract breaks & sweeps from previous agent insights
    const structBreaks = (prevStructure?.insights || [])
      .filter(i => ['BOS', 'CHOCH', 'MSS', 'IDM'].includes(i.type))
      .map(i => ({ type: i.type as any, direction: i.direction!, breakPrice: i.price || 0, breakTimestamp: i.timestamp, breakIndex: i.barIndex || 0, swingPoint: {} as any, structureType: 'SWING' as any, confirmed: true }));

    const sweepInsights = (prevSM?.insights || []).filter(i => i.type === 'LIQUIDITY_SWEEP');

    // LIT Analysis using the module (simplified - builds from available context)
    const litSetups = this.lit.analyze(candles, structBreaks as any, [], [] as any, [], [], []);

    for (const setup of litSetups.slice(-5)) {
      insights.push({
        id: `lit_${++seq}`, agentName: this.name, type: setup.type,
        direction: setup.direction, confidence: setup.confidence,
        price: setup.price, timestamp: setup.timestamp, barIndex: setup.index,
        detail: `${setup.type.replace(/_/g, ' ')} — ${setup.confirmations.length} confirmations`,
        weight: 2.5, tags: [setup.type, ...setup.confirmations.filter(c => c.confirmed).map(c => c.type)],
      });
    }

    // Sweep + Structure break combo = institutional entry signal
    if (sweepInsights.length > 0 && structBreaks.length > 0) {
      const lastSweep = sweepInsights[sweepInsights.length - 1];
      const lastBreak = structBreaks[structBreaks.length - 1];
      if (lastSweep.direction && lastBreak.direction === lastSweep.direction) {
        insights.push({
          id: `lit_${++seq}`, agentName: this.name, type: 'INSTITUTIONAL_ENTRY_SIGNAL',
          direction: lastSweep.direction, confidence: 80, timestamp: Date.now(),
          detail: 'Sweep + Structure break confirmation = institutional entry',
          weight: 3, tags: ['INSTITUTIONAL', 'ENTRY_SIGNAL'],
        });
      }
    }

    const bullish = insights.filter(i => i.direction === 'BULLISH').reduce((s, i) => s + i.weight, 0);
    const bearish = insights.filter(i => i.direction === 'BEARISH').reduce((s, i) => s + i.weight, 0);
    const bias: Bias = bullish > bearish * 1.2 ? 'BULLISH' : bearish > bullish * 1.2 ? 'BEARISH' : 'NEUTRAL';
    const confidence = Math.min(90, litSetups.length > 0 ? litSetups[litSetups.length - 1].confidence : 35);

    return {
      agentName: this.name, status: 'COMPLETE', bias, confidence, insights,
      narrative: `LIT: ${litSetups.length} setups detected. ${insights.length} total insights. Bias: ${bias}.`,
      processingTimeMs: performance.now() - start, timestamp: Date.now(),
    };
  }

  private empty(start: number): AgentOutput {
    return { agentName: this.name, status: 'COMPLETE', bias: 'NEUTRAL', confidence: 30,
      insights: [], narrative: 'Insufficient data for LIT analysis.',
      processingTimeMs: performance.now() - start, timestamp: Date.now() };
  }

  reset(): void { this.lit.reset(); }
}
